k4FWCore/components/UniqueIDGenSvc.cpp

```diff
@@ -18,27 +18,47 @@
  */
 #include "UniqueIDGenSvc.h"
 
-DECLARE_COMPONENT(UniqueIDGenSvc)
+#include <cstdint>
+#include <string>
 
 UniqueIDGenSvc::UniqueIDGenSvc(const std::string& name, ISvcLocator* svcLoc) : base_class(name, svcLoc) {}
 
-StatusCode UniqueIDGenSvc::initialize() {
-  StatusCode sc = Service::initialize();
-  return sc;
-}
-
-const size_t bits32    = std::numeric_limits<uint32_t>::digits;
-const size_t bits64    = std::numeric_limits<uint64_t>::digits;
-const size_t bitsSizeT = std::numeric_limits<size_t>::digits;
+constexpr size_t bits32    = std::numeric_limits<uint32_t>::digits;
+constexpr size_t bits64    = std::numeric_limits<uint64_t>::digits;
+constexpr size_t bitsSizeT = std::numeric_limits<size_t>::digits;
 
 size_t UniqueIDGenSvc::getUniqueID(uint32_t evt_num, uint32_t run_num, const std::string& name) const {
-  std::bitset<bits64>    seed_bits(this->m_seed);
-  std::bitset<bits32>    event_num_bits(evt_num), run_num_bits(run_num);
-  size_t                 str_hash = std::hash<std::string>{}(name);
-  std::bitset<bitsSizeT> name_bits(str_hash);
+  std::bitset<bits64>    seed_bits      = this->m_seed.value();
+  std::bitset<bits32>    event_num_bits = evt_num, run_num_bits = run_num;
+  size_t                 str_hash  = std::hash<std::string>{}(name);
+  std::bitset<bitsSizeT> name_bits = str_hash;
+
+  std::bitset<bits64 + bits32 + bits32 + bitsSizeT> combined_bits;
 
-  std::bitset<bits64 + bits32 + bits32 + bitsSizeT> combined_bits(seed_bits.to_string() + event_num_bits.to_string() +
-                                                                  run_num_bits.to_string() + name_bits.to_string());
+  for (size_t i = 0; i < bits64; i++) {
+    combined_bits[i + bits32 + bits32 + bitsSizeT] = seed_bits[i];
+  }
+  for (size_t i = 0; i < bits32; i++) {
+    combined_bits[i + bits32 + bitsSizeT] = event_num_bits[i];
+  }
+  for (size_t i = 0; i < bits32; i++) {
+    combined_bits[i + bitsSizeT] = run_num_bits[i];
+  }
+  for (size_t i = 0; i < bitsSizeT; i++) {
+    combined_bits[i] = name_bits[i];
+  }
```

Comment:
Is there a reason to set the bits in this semi reversed order? I think it would be more readable if we reverse the order of loops here, because then all the bits would be set in order. Ideally there would be a way to get rid of the loops entirely, but I don't see anything really obvious apart from bit masking which is not really more readable, IMHO.

Reply:
The order keeps unchanged the current IDs, I reversed it now. For masking they all would need to have the same width so the bitsets would need to be converted to the width of the final one; it is probably very similar in terms of what actually happens internally.

Reply:
Does the reversing the order now mean that the IDs are no longer the same?

Reply:
IIUC, the order of the bits (i.e. the resulting bitfield) is still the same. My main complaint was about how the bits are set, because the loops started setting bits from `N - 64` to `N`, then did `N - 64 -32` to `N - 64`, etc.

Reply:
No, the order is the same, the loops are setting the bits from least significant to most significant, the only thing I did after this suggestion was to reverse the order of the loops in the code.

Reply:
Ok, thanks for clarifying!